tests/test_util_worker_pools.py

```diff
@@ -0,0 +1,196 @@
+            pool.join()
+        eq_(1/3.0, pool.success_rate)
+
+
+class MockQueue(Queue):
+    error_count = 0
+
+    def inc_error(self):
+        self.error_count += 1
+
+
+class TestWorker(object):
+
+    def test_factory(self):
+        mock_queue = object()
+        result = Worker.factory(mock_queue)
+        assert isinstance(result, Worker)
+        eq_(mock_queue, result.jobs)
+        eq_(True, result.daemon)
+
+    def test_works_on_callable_job(self):
+        results = list()
+
+        def task():
+            results.append('werk')
+
+        try:
+            q = MockQueue()
+            for i in range(6):
+                q.put(task)
+            rihanna = Worker(q)
+            rihanna.start()
+        finally:
+            q.join()
+
+        eq_(['werk', 'werk', 'werk', 'werk', 'werk', 'werk'], results)
+
+    def test_works_on_job_object(self):
+        results = list()
+
+        original = ['Who Can I * To', '* To You', 'Water *s Dry', '* The World']
+        class MockJob(object):
+            def __init__(self, idx):
+                self.idx = idx
+
+            def run(self):
+                results.append(original[self.idx])
+
+        try:
+            q = MockQueue()
+            for i in range(len(original)):
+                q.put(MockJob(i))
+            rb = Worker(q)
+            rb.start()
+        finally:
+            q.join()
+
+        eq_(sorted(original), sorted(results))
+
+
+class TestDatabaseWorker(DatabaseTest):
+
+    def test_scoped_session(self):
+
+        # Create a mock database object to keep track of methods that
+        # are called.
+        class MockDatabase():
+            def __init__(self, _db):
+                self._expired = 0
+                self._committed = 0
+                self._db = _db
+
+            def expire_all(self):
+                self._expired += 1
+
+            def commit(self):
+                self._committed += 1
+
+            def add(self, item):
+                self._db.add(item)
+
+        # A job that works.
+        def task(_db):
+            identifier = Identifier(type='Keep It', identifier='100')
+            _db.add(identifier)
+
+        # A job that breaks.
+        def broken_task(_db):
+            identifier = Identifier(type='You Can', identifier='Keep It')
+            _db.add(identifier)
+            raise RuntimeError
+
+        try:
+            q = MockQueue()
+            q.put(task)
+            q.put(broken_task)
+
+            mock_db = MockDatabase(self._db)
+            dbw = DatabaseWorker(q, mock_db)
+            dbw.start()
+        finally:
+            q.join()
+
+        # The database was expired for each job, then called one more
+        # time before finding out the queue was empty.
+        eq_(3, mock_db._expired)
+
+        # It was only committed for the working job.
+        eq_(1, mock_db._committed)
+
+        # The DatabaseWorker doesn't roll back the database. It
+        # trusts the task to manage that.
```

Comment:
I think this makes the most sense, given how coverage providers handle the work, but I'm open to other opinions.

Reply:
Is the rationale that coverage providers are responsible for committing, so they are also responsible for rollback if something goes wrong?